common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/MicrosoftAuthorizationRequest.java

```diff
@@ -0,0 +1,279 @@
+        mCorrelationId = correlationId;
+    }
+
+    public PkceChallenge getPkceChallenge() {
+        return mPkceChallenge;
+    }
+
+    public void setPkceChallenge(final PkceChallenge pkceChallenge) {
+        mPkceChallenge = pkceChallenge;
+    }
+
+    public String getExtraQueryParam() {
+        return mExtraQueryParam;
+    }
+
+    public void setExtraQueryParam(final String extraQueryParam) {
+        mExtraQueryParam = extraQueryParam;
+    }
+
+    public String getLibraryVersion() {
+        return mLibraryVersion;
+    }
+
+    public void setLibraryVersion(final String libraryVersion) {
+        mLibraryVersion = libraryVersion;
+    }
+
+    /**
+     * Convert the query parameters string to pair and add into the request parameters map.
+     */
+    protected void appendExtraQueryParameters(final String queryParams, final Map<String, String> requestParams) throws ClientException {
+        final Map<String, String> extraQps = constructQueryParamsMap(queryParams);
+        final Set<Map.Entry<String, String>> extraQpEntries = extraQps.entrySet();
+        for (final Map.Entry<String, String> extraQpEntry : extraQpEntries) {
+            if (requestParams.containsKey(extraQpEntry.getKey())) {
+                throw new ClientException(ErrorStrings.DUPLICATE_QUERY_PARAMETER,
+                        "Extra query parameter " + extraQpEntry.getKey() + " is already sent by "
+                                + "the SDK. ");
+            }
+
+            requestParams.put(extraQpEntry.getKey(), extraQpEntry.getValue());
+        }
+    }
+
+    public String generateEncodedState() throws UnsupportedEncodingException {
+        final UUID stateUUID1 = UUID.randomUUID();
+        final UUID stateUUID2 = UUID.randomUUID();
+        final String state = stateUUID1.toString() + "-" + stateUUID2.toString();
+        return Base64.encodeToString(state.getBytes("UTF-8"), Base64.NO_PADDING | Base64.URL_SAFE);
+    }
+
+    public String decodeState(final String encodedState) {
+        if (StringUtil.isEmpty(encodedState)) {
+            Logger.warn(TAG, "Decode state failed because the input state is empty.");
+            return null;
+        }
+
+        final byte[] stateBytes = Base64.decode(encodedState, Base64.NO_PADDING | Base64.URL_SAFE);
+        return new String(stateBytes, Charset.defaultCharset());
+    }
+
+
+    protected void addExtraQueryParameter(final String key, final String value, final Map<String, String> requestParams) {
+        if (!isNullOrBlank(key) && !isNullOrBlank(value)) {
+            requestParams.put(key, value);
+        }
+    }
+
+
+    /**
+     * Sanitizes the query portion of a URL and convert the query parameters into a map.
+     *
+     * @param url The url to decode for.
+     * @return A map of all the query parameter pairs in the URL's query portion.
+     */
+    private static Map<String, String> constructQueryParamsMap(final String url) {
```

Comment:
This implementation makes a lot of assumptions about the formatting of query params....

Consider the following _valid_ URL and how this code would treat it:
`http://foo.bar?list_a=1&list_a=2&list_a=3&list_b[]=1&list_b[]=2&list_b[]=3&list_c=1,2,3`